### PR TITLE
n_distinct() can use .data in mutate()

### PR DESCRIPTION
**Pull request by krlmlr**
Closes tidyverse/tibble#721.

### CODE REVIEW

File: R/distinct.R [hunk 1]
```diff
@@ -128,15 +128,15 @@ distinct.data.frame <- function(.data, ..., .keep_all = FALSE) {
 #'
 #' This is a faster and more concise equivalent of `length(unique(x))`
 #'
-#' @param \dots vectors of values
+#' @param \dots Vectors of values, passed on to [list2()]
```
Comment (lionel-) [inline R/distinct.R]: Should use dynamic dots tag.

**Comment by krlmlr** [comment]
I checked, no other calls of the form `tibble(...)` .

I'm happy to revise the `tibble()` UI, let's discuss in https://github.com/tidyverse/tibble/issues/721.

**Comment by lionel-** [comment]
> I supposed !!!list2() would no longer be needed after this : tidyverse/tibble@d0707ef
>
> Perhaps we can leave a comment to change then.

Going through `list()` or `list2()` is definitely needed, otherwise `n_distinct()` would inherit the semantics of `tibble()`.

**Comment by krlmlr** [comment]
In other words, do we allow `mutate()` semantics in `n_distinct()`? f8aa4561, not part of this branch yet.

**Comment by lionel-** [comment]
IMO it would only make the function more confusing. However it's probably not worth to spend time thinking about this. Tibble semantics were introduced by error, and it seems better to avoid changing semantics without an actual need or request.

**Comment by DavisVaughan** [comment]
FWIW I think the simplest implementation is:

```r
n_distinct <- function(..., na.rm = FALSE) {
  data <- list2(...)
  names(data) <- as.character(seq_along(data))
  
  size <- vec_size_common(!!!data)
  data <- vec_recycle_common(!!!data, .size = size)
  
  data <- new_data_frame(data, n = size)
  
  if (isTRUE(na.rm)) {
    data <- vec_slice(data, !vec_equal_na(data))
  }
  
  vec_unique_count(data)
}
```

(This is on top of #4950, which tweaks the `na.rm = TRUE` case)

This:

- Avoids `tibble()` data mask issues (which were also fixed by https://github.com/tidyverse/tibble/pull/725 anyways)
- Does not have `mutate()` semantics

**Comment by hadley** [comment]
@DavisVaughan do you want to alter #4950 to do that? I think that's the simplest way forward.

**Pull request closed without merging**